validmind/unit_metrics/__init__.py

```diff
@@ -0,0 +1,192 @@
+import pandas as pd
+import json
+import hashlib
+import importlib
+
+from ..utils import get_model_info
+
+
+global_metric_values = {}
+
+def _serialize_params(params):
+    """
+    Serialize the parameters to a unique hash.
+
+    This function serializes the parameters dictionary to a JSON string,
+    then creates a SHA-256 hash of the string to ensure a unique identifier
+    for the parameters.
+
+    Args:
+        params (dict): The parameters to be serialized.
+
+    Returns:
+        str: A SHA-256 hash of the JSON string representation of the params.
+    """
+    # Convert the params to a JSON string, ensuring consistent ordering
+    params_json = json.dumps(params, sort_keys=True)
+
+    # Create a SHA-256 hash of the JSON string
+    hash_object = hashlib.sha256(params_json.encode())
+    return hash_object.hexdigest()
+
+def _serialize_model(model):
+    """
+    Generate a SHA-256 hash for a scikit-learn model based on its type and parameters.
+
+    Args:
+        model VMModel: The model to be serialized.
+
+    Returns:
+        str: A SHA-256 hash of the model's description.
+    """
+
+    if model is None:
+        # You can return a predefined hash, or handle it in another appropriate way
+        model_info = "none_model_hash"
+    else:
+        model_info = get_model_info(model)
+
+    model_json = json.dumps(model_info, sort_keys=True)
+
+    # Create a SHA-256 hash of the JSON string
+    hash_object = hashlib.sha256(model_json.encode())
+    return hash_object.hexdigest()
+
+def _serialize_dataset(dataset):
+    """
+    Serialize the description of the dataset input to a unique hash.
+
+    This function generates descriptive statistics of the dataset, converts
+    the description to a JSON string, and then creates a SHA-256 hash of the string
+    to ensure a unique identifier for the dataset's description.
+
+    Args:
+        dataset (pd.DataFrame): The dataset whose description is to be serialized.
+
+    Returns:
+        str: A SHA-256 hash of the JSON string representation of the dataset's description.
+    """
+    if isinstance(dataset._df, pd.DataFrame):
```

Comment:
Any reason this check is needed? We expect `init_dataset` will do all the plumbing required for this to work.